pandas/core/algorithms.py

```diff
@@ -1991,13 +1993,14 @@ def union_with_duplicates(lvals: ArrayLike, rvals: ArrayLike) -> ArrayLike:
     -----
     Caller is responsible for ensuring lvals.dtype == rvals.dtype.
     """
-    indexer = []
     l_count = value_counts(lvals, dropna=False)
     r_count = value_counts(rvals, dropna=False)
     l_count, r_count = l_count.align(r_count, fill_value=0)
-    unique_array = unique(concat_compat([lvals, rvals]))
-    unique_array = ensure_wrapped_if_datetimelike(unique_array)
-
-    for i, value in enumerate(unique_array):
-        indexer += [i] * int(max(l_count.at[value], r_count.at[value]))
-    return unique_array.take(indexer)
+    final_count = np.maximum(l_count, r_count).astype("int", copy=False)
```

Comment:
You've got 2 MultiIndex checks now. I'd rather handle this on the _union level. Is there any performance benefit of doing it like this?

Additionally, I think you can use unique_vals.take(repeats)
That's more in line with what we are doing elsewhere

Reply:
I think we need the MultiIndex checks within union_with_duplicates, otherwise we lose dtypes in there. I'm not sure how how this could be done only on the _union level. Let me know if you have any suggestions. 

The repeat logic is not quite a simple `take`:

```
np.repeat(unique_vals, repeats)
```

is equivalent to:

```
indexer = np.arange(len(unique_vals))
indexer = np.repeat(indexer, repeats)
unique_vals.take(indexer)
```

Reply:
You could simply return ``final_count`` and handle the rest In the _union level. Alternatively, you could  just pass in self and other and handle everything on the lower level. Or add another argument ``unique_vals`` to the function

Reply:
Thx, makes sense. I pushed the type checking down into union_with_duplicates.